### ISSUE TITLE
Polish & release 3.1

### ISSUE THREAD

**Issue by Pathoschild** [Pathoschild/FluentHttpClient#66]
This is just a tracking ticket to finalise the 3.1 release:

* [x] Ensure all unit tests pass.
* [x] Ensure code conventions are applied, all code is documented, etc.
* [x] Polish documentation, release notes, etc.
* [x] Prepare NuGet package.
* [x] Test the client in a few projects.
* [x] Deploy to NuGet.

Feel free to comment if there's anything else you want to include in the 3.1 release.

**Comment by Pathoschild** [comment]
@Jericho Here's the [3.1 NuGet package](https://github.com/Pathoschild/FluentHttpClient/files/1291092/Pathoschild.Http.FluentClient.3.1.0.nupkg.zip). It was created with the [new MSBuild/NuGet integration](https://blog.nuget.org/20170316/NuGet-now-fully-integrated-into-MSBuild.html); let me know if you run into any problems using it. I'll release soon if nothing comes up.

**Comment by Jericho** [comment]
I get the following exception when referencing this nuget package: `Package restore failed. Rolling back package changes`.

This message is frustratingly vague though.

**Comment by Pathoschild** [comment]
Are there more details in the output window (possibly under build or package manager)?

**Comment by Jericho** [comment]
I was able to get more details about this error:
```
Package Pathoschild.Http.FluentClient 3.1.0 is not compatible with net452 (.NETFramework,Version=v4.5.2). Package Pathoschild.Http.FluentClient 3.1.0 supports: netstandard2.0 (.NETStandard,Version=v2.0)	
Package Pathoschild.Http.FluentClient 3.1.0 is not compatible with netstandard1.3 (.NETStandard,Version=v1.3). Package Pathoschild.Http.FluentClient 3.1.0 supports: netstandard2.0 (.NETStandard,Version=v2.0)
```

**Comment by Pathoschild** [comment]
That's pretty weird. It's definitely not targeting .NET Standard 2.0 (I haven't even updated VS to support that version yet):
> ![](https://user-images.githubusercontent.com/230581/30255695-7cf30826-9673-11e7-9be1-d39756e63234.png)
> ![](https://user-images.githubusercontent.com/230581/30255615-03867be4-9673-11e7-9202-8416097efd7a.png)

I'll do some digging to see if I made a mistake somewhere, though it's working fine in my projects. 🤔

**Comment by Jericho** [comment]
I removed FluentClient from my solution, closed VS.NET, cleared my local nuget cache, reopened VS.NET and attempted to add reference to FluentClient and this time I got an error message about version mismatch between Json.Net in the FluentClient package and my project. So I upgraded Json.net in my project to match the version referenced by FluentClient and attempted to add reference to FluentClient. This time, it worked.

However, my project will no longer compile because one of my classes is not implementing an new attribute that was added to the `IRetryConfig` interface: `'SendGridRetryStrategy' does not implement interface member 'IRetryConfig.RetryOnTimeout'`.

Adding this new property to my class is easy enough but I did not expect such a breaking change in a minor release. Did you intend to introduce a breaking change in this release? If so, the major version should be increased.

**Comment by Pathoschild** [comment]
Oh, I didn't account for custom implementations of that interface. I don't want to make this a 4.0 release, since it's a pretty small release. Maybe we can make retrying timeouts automatic, and defer the config option for an upcoming 4.0 release?

**Comment by Jericho** [comment]
I was curious why you decided to treat timeouts differently than any other response? Why not simply delegate the decision whether to retry them or not to `IRetryConfig.ShouldRetry` and therefore allow developers to decide for themselves whether they want timeouts to be retried?

**Comment by Jericho** [comment]
I think you set the `response` to `null` when a timeout occurs, right? If so, my custom retry strategy would look something like this:

```
public bool ShouldRetry(HttpResponseMessage response)
{
    // Timeout
    if (response == null) return true;

    // TOO MANY REQUESTS
    elseif (response.StatusCode == (HttpStatusCode)429) return true;

    // Any other case
    else return false;
}
````

**Comment by Jericho** [comment]
One more comment: with the current logic, I find it weird that you don't invoke `IRetryConfig.ShouldRetry` to decide if the timeout should be retried or not and yet you do invoke `IRetryConfig.GetDlay` in order to get the TimeSpan for the delay task. It's not clear to me why you decided to invoke one but not the other when a timeout occurs.

**Comment by Jericho** [comment]
In other words, I would change this line:
```
bool shouldRetry = response != null ? this.Config.ShouldRetry(response) : this.Config.RetryOnTimeout;
```
with this one:
```
bool shouldRetry = this.Config.ShouldRetry(response);
```

**Comment by Pathoschild** [comment]
Reusing `ShouldRetry` for timeouts would be ideal, but there's no `HttpResponseMessage` to check since HttpClient throws a `TaskCancelledException` on timeout. Just passing `null` would be unintuitive and a breaking change (since existing implementations would throw null reference exceptions).

Maybe we could create a fake response with a relevant status code, such as...

status               | note
-------------------- | ---------------------------
408 Request Timeout | Makes intuitive sense, though not technically applicable (it's supposed to mean the server timed out while waiting for the client).
508 Gateway Timeout | Kind of applicable, if you consider the HTTP client a gateway/proxy.
598 Network Read Timeout | This is the closest match, though it's unofficial and there's no `HttpStatusCode` constant for it.

**Comment by Jericho** [comment]
Not having a HttpStatusCode constant is not a big problem (at least not for me). For instance, one of the APIs I use returns HTTP 429 when I exceed their throttling limit and there is no constant for this code either. That's why my retry logic checks for: ` (HttpStatusCode)429`

You can even create your own constants if you want. Like so:
```
public bool ShouldRetry(HttpResponseMessage response)
{
	const HttpStatusCode TOO_MANY_REQUESTS = (HttpStatusCode)429;
	const HttpStatusCode TIMEOUT = (HttpStatusCode)598;

	// Should never happen but let's be safe
	if (response == null) return false;

	// Timeout
	else if (response.StatusCode == TIMEOUT) return true;

	// Exceeded throttling limit
	else if (response.StatusCode == TOO_MANY_REQUESTS) return true;

	// Any other case
	else return false;
}
```

**Comment by Pathoschild** [comment]
Yep, that's what I was thinking if we use HTTP 598.

**Comment by Jericho** [comment]
One more thing I wanted to bring up: are you sure `catch (TaskCanceledException) when (...)` will only catch timeout exceptions? Seems to me that you would get similar `TaskCanceledException` if, for example, another async task cancels the CancellationToken.

**Comment by Jericho** [comment]
Did some research regarding a task canceled due to the CancellationToken being cancelled vs canceled due to timeout and found a [question on StackOverflow](https://stackoverflow.com/questions/12666922/distinguish-timeout-from-user-cancellation) (written in 2012) on this very topic. Someone suggested (in 2013) checking the exception's `CancellationToken.IsCancellationRequested` property to distinguish between these two but someone else (in 2015) points out that there's an issue in HttpClient and the `IsCancellationRequested` cannot be relied upon. There's also [another question](https://stackoverflow.com/questions/29319086/cancelling-an-httpclient-request-why-is-taskcanceledexception-cancellationtoke) where someone suggests a unit test to demonstrate the issue and a lengthy explanation.

I ran the unit test in the SO question and it completes successfully in both net452 and netstandard. Seems like this issue has been resolved in the System.Net.Http nuget package (probably in version 4.0.0).

Here's the unit test in question (in the SO article, they mention that the "Assert" in the 'catch' fails but it appears to be working fine now):
```
[Test(Description = "Verify that we can detect that a HTTP request has been voluntarily canceled versus cancelled by a timeout.")]
public async Task Verify_suggestion_from_StackOverflow_to_determine_http_request_timeout()
{
    var cts = new CancellationTokenSource();

    try
    {
        // get a "hot" task
        var task = new HttpClient().GetAsync("http://www.google.com", cts.Token);

        // request cancellation
        cts.Cancel();

        await task;

        Assert.Fail("expected TaskCanceledException to be thrown");
    }
    catch (TaskCanceledException ex)
    {
        Assert.IsTrue(ex.CancellationToken.IsCancellationRequested, "expected cancellation requested on token attached to exception");
    }
}
```

So, long story short: I propose we improve the RetryCoordinator like so:
```
try
{
    response = await dispatcher(request).ConfigureAwait(false);
}
catch (TaskCanceledException ex) when (!ex.CancellationToken.IsCancellationRequested)
{
    response = request.Message.CreateResponse(this.TimeoutStatusCode);
}
```

**Comment by Pathoschild** [comment]
Thanks! I found the same info. The bug might still happen in some versions of HttpClient, but we have the original token anyway:
```cs
catch (TaskCanceledException) when (!request.CancellationToken.IsCancellationRequested)
```

I made that change and added a unit test for it, and replaced `RetryOnTimeout` with the HTTP 598 response as discussed above. Here's a fresh package:
* [Pathoschild.Http.FluentClient.3.1.0.nupkg](https://github.com/Pathoschild/FluentHttpClient/files/1294325/Pathoschild.Http.FluentClient.3.1.0.nupkg.zip)

**Comment by Jericho** [comment]
Looks very good. I simply changed `maxRetries: 5` to `maxRetries: 4` in my code and all my unit tests run successfully.

**Comment by Jericho** [comment]
When do you plan to release 3.1?

**Comment by Pathoschild** [comment]
I'm aiming for next week. I'm travelling this weekend, and I want to test-migrate one of my projects to the new version before release.

**Comment by Pathoschild** [comment]
3.1 is live on NuGet. 🎉
